File: android/src/main/java/app/capgo/gtm/GoogleTagManagerPlugin.java
package app.capgo.gtm;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.util.Map;

@CapacitorPlugin(name = "GoogleTagManager")
public class GoogleTagManagerPlugin extends Plugin {

    private GoogleTagManager implementation;

    @Override
    public void load() {
        implementation = new GoogleTagManager(getContext());
    }

    @PluginMethod
    public void initialize(PluginCall call) {
        String containerId = call.getString("containerId");
        if (containerId == null) {
            call.reject("Missing containerId parameter");
            return;
        }

        Double timeout = call.getDouble("timeout");

        implementation.initialize(
            containerId,
            timeout,
            new GoogleTagManager.Callback() {
                @Override
                public void onSuccess() {
                    call.resolve();
                }

                @Override
                public void onFailure(String error) {
                    call.reject(error);
                }
            }
        );
    }

    @PluginMethod
    public void push(PluginCall call) {
        String event = call.getString("event");
        if (event == null) {
            call.reject("Missing event parameter");
            return;
        }

        JSObject parameters = call.getObject("parameters");
        Map<String, Object> paramMap = parameters != null ? GoogleTagManager.jsObjectToMap(parameters) : null;

        implementation.push(
            event,
            paramMap,
            new GoogleTagManager.Callback() {
                @Override
                public void onSuccess() {
                    call.resolve();
                }

                @Override
                public void onFailure(String error) {
                    call.reject(error);
                }
            }
        );
    }

    @PluginMethod
    public void setUserProperty(PluginCall call) {
        String key = call.getString("key");
        Object value = call.getData().opt("value");

        if (key == null || value == null) {
            call.reject("Missing key or value parameter");
            return;
        }

        implementation.setUserProperty(
            key,
            value,
            new GoogleTagManager.Callback() {
                @Override
                public void onSuccess() {
                    call.resolve();
                }

                @Override
                public void onFailure(String error) {
                    call.reject(error);
                }
            }
        );
    }

    @PluginMethod
    public void getValue(PluginCall call) {
        String key = call.getString("key");
        if (key == null) {
            call.reject("Missing key parameter");
            return;
        }

        implementation.getValue(
            key,
            new GoogleTagManager.ValueCallback() {
                @Override
                public void onSuccess(Object value) {
                    JSObject ret = new JSObject();
                    ret.put("value", value);
                    call.resolve(ret);
                }

                @Override
                public void onFailure(String error) {
                    call.reject(error);
                }
            }
        );
    }

    @PluginMethod
    public void reset(PluginCall call) {
        implementation.reset(
            new GoogleTagManager.Callback() {
                @Override
                public void onSuccess() {
                    call.resolve();
                }

                @Override
                public void onFailure(String error) {
                    call.reject(error);
                }
            }
        );
    }
}

File: ios/Sources/GoogleTagManagerPlugin/GoogleTagManagerPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(GoogleTagManagerPlugin)
public class GoogleTagManagerPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "GoogleTagManagerPlugin"
    public let jsName = "GoogleTagManager"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "initialize", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "push", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setUserProperty", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getValue", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "reset", returnType: CAPPluginReturnPromise)
    ]
    private let implementation = GTMManager()

    @objc func initialize(_ call: CAPPluginCall) {
        guard let containerId = call.getString("containerId") else {
            call.reject("Missing containerId parameter")
            return
        }
        
        let timeout = call.getDouble("timeout")
        
        implementation.initialize(containerId: containerId, timeout: timeout) { success, error in
            if success {
                call.resolve()
            } else {
                call.reject(error?.localizedDescription ?? "Failed to initialize Google Tag Manager")
            }
        }
    }

    @objc func push(_ call: CAPPluginCall) {
        guard let event = call.getString("event") else {
            call.reject("Missing event parameter")
            return
        }
        
        let parameters = call.getObject("parameters") ?? [:]
        
        implementation.push(event: event, parameters: parameters) { success, error in
            if success {
                call.resolve()
            } else {
                call.reject(error?.localizedDescription ?? "Failed to push event")
            }
        }
    }

    @objc func setUserProperty(_ call: CAPPluginCall) {
        guard let key = call.getString("key") else {
            call.reject("Missing key parameter")
            return
        }
        
        guard let value = call.getValue("value") else {
            call.reject("Missing value parameter")
            return
        }
        
        implementation.setUserProperty(key: key, value: value) { success, error in
            if success {
                call.resolve()
            } else {
                call.reject(error?.localizedDescription ?? "Failed to set user property")
            }
        }
    }

    @objc func getValue(_ call: CAPPluginCall) {
        guard let key = call.getString("key") else {
            call.reject("Missing key parameter")
            return
        }
        
        implementation.getValue(key: key) { value, error in
            if let error = error {
                call.reject(error.localizedDescription)
            } else {
                call.resolve(["value": value as Any])
            }
        }
    }

    @objc func reset(_ call: CAPPluginCall) {
        implementation.reset() { success, error in
            if success {
                call.resolve()
            } else {
                call.reject(error?.localizedDescription ?? "Failed to reset")
            }
        }
    }
}
